### PR TITLE
Remove default empty state of `sf::Texture`

### PR DESCRIPTION
**Pull request by ChrisThrasher**
## Description

Next in line after #3025 and others before it.

### CODE REVIEW

File: include/SFML/Graphics/Texture.hpp
Comment (vittorioromeo) [inline include/SFML/Graphics/Texture.hpp:96]:
```suggestion
    /// \return Texture if loading was successful, `std::nullopt` otherwise
```

File: include/SFML/Graphics/Texture.hpp
Comment (vittorioromeo) [inline include/SFML/Graphics/Texture.hpp:116]:
```suggestion
    /// \return Texture if loading was successful, `std::nullopt` otherwise
```

File: include/SFML/Graphics/Texture.hpp
Comment (vittorioromeo) [inline include/SFML/Graphics/Texture.hpp:140]:
```suggestion
    /// \return Texture if loading was successful, `std::nullopt` otherwise
```

File: include/SFML/Graphics/Texture.hpp
Comment (vittorioromeo) [inline include/SFML/Graphics/Texture.hpp:122]:
```suggestion
                                                             const IntRect&               area = {});
```

File: include/SFML/Graphics/Texture.hpp
Comment (vittorioromeo) [inline include/SFML/Graphics/Texture.hpp:147]:
```suggestion
                                                               const IntRect& area = {});
```

File: include/SFML/Graphics/Texture.hpp
Comment (vittorioromeo) [inline include/SFML/Graphics/Texture.hpp:164]:
```suggestion
    /// \return Texture if loading was successful, `std::nullopt` otherwise
```

File: include/SFML/Graphics/Texture.hpp
Comment (vittorioromeo) [inline include/SFML/Graphics/Texture.hpp:169]:
```suggestion
    [[nodiscard]] static std::optional<Texture> loadFromStream(InputStream& stream, const IntRect& area = {});
```

File: include/SFML/Graphics/Texture.hpp
Comment (vittorioromeo) [inline include/SFML/Graphics/Texture.hpp:186]:
```suggestion
    /// \return Texture if loading was successful, `std::nullopt` otherwise
```

File: include/SFML/Graphics/Texture.hpp
Comment (vittorioromeo) [inline include/SFML/Graphics/Texture.hpp:191]:
```suggestion
    [[nodiscard]] static std::optional<Texture> loadFromImage(const Image& image, const IntRect& area = {});
```

File: src/SFML/Graphics/Texture.cpp
Comment (vittorioromeo) [inline src/SFML/Graphics/Texture.cpp:188]:
```suggestion
```

This check is useless now, right?

File: src/SFML/Graphics/Texture.cpp
Comment (vittorioromeo) [inline src/SFML/Graphics/Texture.cpp:201]: Ditto for the other checks -- unless I am missing something, they should all have their default value, as this is a brand new texture. Right?

**Review by vittorioromeo — DISMISSED** [review]

File: include/SFML/Graphics/Texture.hpp
Comment (ChrisThrasher) [reply include/SFML/Graphics/Texture.hpp:116]: What makes you feel strongly about this order of words? There are lots of SFML APIs that use slightly different variations of this same sentence so there is no one standard phrasing for me to pick from. Because there is little consistency I don't put too much thought into it. Perhaps it would be worthwhile to go through all our docs and unify our phrasing for APIs with the same semantics.

File: src/SFML/Graphics/Texture.cpp
Comment (ChrisThrasher) [reply src/SFML/Graphics/Texture.cpp:185]: `sf::RenderTexture` takes advantage of this default constructor. To remove `Texture::Texture()` would result in more downstream changes to `RenderTexture`. However, if we also change `RenderTexture::create` to return a `std::optional<RenderTexture>` then this problem isn't so bad. I wonder if this PR should be postponed until I can perform this refactor inside `RenderTexture`. That order of operations may make more sense.

File: include/SFML/Graphics/Texture.hpp
Comment (vittorioromeo) [reply include/SFML/Graphics/Texture.hpp:116]: I don't feel strongly about it, it just sounds better in my mind. Feel free to ignore and I agree with your last sentence

File: src/SFML/Graphics/Font.cpp
Comment (ChrisThrasher) [reply src/SFML/Graphics/Font.cpp:757]: I thought this was a C++20 feature. I suppose I'll remove it and wait for CI to yell at me if it is in fact a C++20 feature.

File: src/SFML/Graphics/Texture.cpp
Comment (ChrisThrasher) [reply src/SFML/Graphics/Texture.cpp:188]: Yeah I think you're right. Good catch.

File: src/SFML/Graphics/Texture.cpp
Comment (ChrisThrasher) [reply src/SFML/Graphics/Texture.cpp:201]: `m_isRepeated` defaults to `false` and is never touched prior to this line so yeah that will always evaluate to `true` and thus can be removed.

I think this whole section can be removed as well.
```cpp
    if (texture.m_sRgb && !textureSrgb)
    {
        static bool warned = false;

        if (!warned)
        {
#ifndef SFML_OPENGL_ES
            err() << "OpenGL extension EXT_texture_sRGB unavailable" << '\n';
#else
            err() << "OpenGL ES extension EXT_sRGB unavailable" << '\n';
#endif
            err() << "Automatic sRGB to linear conversion disabled" << std::endl;

            warned = true;
        }

        texture.m_sRgb = false;
    }
    ```
`m_sRgb` defaults to `false` and is never touched prior to this line so I don't think there is ever a way to enter this conditional. This is a good sign because it shows that this API change systematically eliminates certain errors that were made possible by the previous API.

Comment (ChrisThrasher) [reply src/SFML/Graphics/Texture.cpp:201]: As I look closer, `Texture::setSrgb` is only useful if you call it prior to loading the texture. I think that needs to be added as a function parameter to ensure there is still a way to set this. I believe the same can be said for `Texture::setSmooth` and `Texture::setRepeated`.

**Comment by ChrisThrasher** [comment]
I discovered a big issue with the current implementation. `sf::Texture::setSrgb`, `sf::Texture::setSmooth`, and `sf::Texture::setRepeated` are effectively used as constructor parameters and thus need to be reflected as parameters of the `create()` function. The definitions of those functions need to be inlined into `create()` as well so that functionality is not lost.

**Comment by ChrisThrasher** [comment]
> I discovered a big issue with the current implementation. `sf::Texture::setSrgb`, `sf::Texture::setSmooth`, and `sf::Texture::setRepeated` are effectively used as constructor parameters and thus need to be reflected as parameters of the `create()` function.

I fixed this by adding parameters to all four static factory functions for setting the smooth, sRbg, and repeated properties. I believe this fixes the functionality issues. Now you can construct a texture with those settings set to something non-default.

The new problem is that an API with 3 consecutive `bool` parameters _sucks_. We need a better solution. I'm thinking it may be best to create a `sf::Texture::Settings` struct with 3 fields for these 3 parameters. Then we can use a default parameter to set this so that users who don't care don't have to write it out. See all the `false, false, false` I had to add hurts, but I'm sure a better solution is possible.

File: include/SFML/Graphics/Texture.hpp
Comment (texus) [inline include/SFML/Graphics/Texture.hpp:693]:
```suggestion
/// const auto texture = sf::Texture::loadFromFile("texture.png").value();
```

**Comment by ChrisThrasher** [comment]
I need to take some more time with this PR so I'm marking it as a draft to indicate that it's not ready for review in its current state.

**Comment by ChrisThrasher** [comment]
I took another shot at this PR. This time I removed `sf::Texture::setSrgb` since it seemingly had no more purpose as a member function. The only place that value is read is within the creation factory function. Setting that value after creation would do nothing so it seemed bug-prone to leave it there. Thus I added 1 parameter to the texture factory functions for specifying sRgb or not so this functionality is not lost.

File: test/Graphics/Texture.test.cpp
Comment (ChrisThrasher) [inline test/Graphics/Texture.test.cpp:305]: Something to know about `setSrgb()` prior to this PR is that it must be called before `create()` for it to work. In this case it was called after `create()` meaning the texture didn't actually attempt to enable SRGB.

My API refactor means this parameter must be provided upon creation. As it turns out this feature doesn't work on Windows OpenGL ES so the `create()` function automatically disables it and emits a warning. That means we can't test for `isSrgb` being `true` since on that one config it's false.

File: src/SFML/Graphics/Font.cpp
Comment (ChrisThrasher) [reply src/SFML/Graphics/Font.cpp:779]: I kept this in my latest revision of this PR. I don't like what I wrote but can't picture anything better at the moment.

Comment (kimci86) [reply src/SFML/Graphics/Font.cpp:779]: The natural next step would be that Font::Page is created with a factory function.

File: test/Graphics/Texture.test.cpp
Comment (kimci86) [reply test/Graphics/Texture.test.cpp:305]: A comment explaining this in the code would be valuable so it doesn't look like testing it was just overlooked.

File: src/SFML/Graphics/Font.cpp
Comment (ChrisThrasher) [reply src/SFML/Graphics/Font.cpp]: I went ahead and made this change because it seems sensible. What happened is now the null check occurs downstream in `sf::Font::loadPage` where the pages are actually constructed. If construction fails inside `loadPage` then what do we do? I chose to `assert` that loading succeeds but we could just as well use `.value()` in which case we're back to the same solution as what I previously wrote. 

Should we go so far as to make `sf::Font::loadPage` return something nullable and put the onus users? That's possible but I don't love the idea.

**Comment by ChrisThrasher** [comment]
Rebased on `master` and fixed merge conflicts.

File: examples/android/app/src/main/jni/main.cpp
Comment (vittorioromeo) [inline examples/android/app/src/main/jni/main.cpp:89]: This changes the behavior to throwing an exception rather than returning `EXIT_FAILURE`. Given that some behavior changes ended up being reverted in #3051, do we want to maintain the same behavior as before here too?

File: examples/cocoa/CocoaAppDelegate.mm
Comment (vittorioromeo) [inline examples/cocoa/CocoaAppDelegate.mm:57]: Ditto about behavioral changes, we also lose the `NSLog` call here. Note that we already changed behavior in line 55 for the font, so I'm not sure...

File: examples/opengl/OpenGL.cpp
Comment (vittorioromeo) [inline examples/opengl/OpenGL.cpp:61]: Not going to repeat myself for all the other ones, but all of these `.value()` calls in the examples change behavior and such changes were reverted in #3051 IIRC.

Maybe we should have a policy about throwing in examples?

File: include/SFML/Graphics/Font.hpp
Comment (vittorioromeo) [inline include/SFML/Graphics/Font.hpp:319]:
```suggestion
        explicit Page(Texture&& texture);
```

**Review by vittorioromeo — CHANGES_REQUESTED** [review]
Minor considerations but looks good to me otherwise, happy to approve once you take a look and decide what to do.

File: examples/android/app/src/main/jni/main.cpp
Comment (ChrisThrasher) [reply examples/android/app/src/main/jni/main.cpp:89]: > This changes the behavior to throwing an exception

In practice the result is the same. Failure to load means the program stops immediately. 

> Given that some behavior changes ended up being reverted in https://github.com/SFML/SFML/pull/3051, do we want to maintain the same behavior as before here too?

#3051 was different. In that case I truly did change behavior in a way that totally broke the example on certain platforms. For now I want to use the same `.value()` pattern in our examples until we come to a consensus on a better alternative. I'm not sure exactly what alternative you or others may prefer.

File: examples/cocoa/CocoaAppDelegate.mm
Comment (ChrisThrasher) [reply examples/cocoa/CocoaAppDelegate.mm:57]: To be honest my attitude here was "meh whatever" about removing those logs. I'm not married to this solution and can go back to the previous solution if you feel strongly about keeping that.

Comment (vittorioromeo) [reply examples/cocoa/CocoaAppDelegate.mm:57]: I never used/targeted Cocoa, I have no idea if those logs provide any real value or not, but honestly your solution is simpler and probably good enough. Unless you know anyone who targets Cocoa that can provide more feedback, I don't feel strongly about keeping the logs at all.

File: examples/android/app/src/main/jni/main.cpp
Comment (vittorioromeo) [reply examples/android/app/src/main/jni/main.cpp:89]: Makes sense, you are correct that #3051 was a special case. Still, would be nice to have a policy regarding exceptions in our examples. Resolving for now.

**Review by vittorioromeo — APPROVED** [review]

File: src/SFML/Graphics/Font.cpp
Comment (eXpl0it3r) [inline src/SFML/Graphics/Font.cpp:763]: Any reason we didn't use `create` as with most other factory functions?

Comment (ChrisThrasher) [reply src/SFML/Graphics/Font.cpp:763]: [Good point](https://github.com/SFML/SFML/pull/3103)

**Pull request merged**
